# Week 3: Python Basics (Part II) - Class Exercises

We first load the packages and data necessary for this exercise.

In [6]:
import pandas as pd
import numpy as np
from pydataset import data

# Load the airquality dataset
airquality = data('airquality')

---

## 1. Lottery Numbers

In order to win the top prize in a particular lottery, one must match all 6 numbers between 1 and 49 drawn by the lottery organizer.

Using the `random.choice()` function from the `numpy` library, generate **six distinct** integers between 1 and 49 (inclusive) and store the results in an object named `lottery_num`. 

Ensure the numbers do not contain any duplicates. Print the numbers in **ascending order**.

Your code should produce something like

The lottery numbers are: [14 18 28 45 46 48].

In [2]:
%pip install numpy
import numpy as np

lottery_num = np.random.choice(range(1, 50), size=6, replace=False)
lottery_num = np.sort(lottery_num)

print(f"The lottery numbers are: {lottery_num}")



[notice] A new release of pip is available: 24.3.1 -> 25.3
[notice] To update, run: python.exe -m pip install --upgrade pip


  Using cached numpy-2.4.1-cp312-cp312-win_amd64.whl.metadata (6.6 kB)
Using cached numpy-2.4.1-cp312-cp312-win_amd64.whl (12.3 MB)
Note: you may need to restart the kernel to use updated packages.
The lottery numbers are: [ 2  7 10 15 21 32]


---

## 2. Air Quality in New York City

In this question, we will examine daily air quality in New York City in 1973. 

The data are available through the `pydataset` package, which we installed with `requirements.txt`.

### (a) Determine the shape of the dataset and the number of unique months.
**Hint:** Have a look at `nunique` for the number of unique months.

In [4]:
%pip install pydataset
from pydataset import data
import pandas as pd

# Load dataset
df = data("airquality")

# Display the first few rows
df.head()
# (a) Shape of dataset
print("Shape of dataset:", df.shape)

# Number of unique months
print("Number of unique months:", df["Month"].nunique())


  Using cached pydataset-0.2.0-py3-none-any.whl
  Using cached tzdata-2025.3-py2.py3-none-any.whl.metadata (1.4 kB)
   ---------------------------------------- 0.0/9.7 MB ? eta -:--:--
   ---------------------------------------- 0.0/9.7 MB ? eta -:--:--
   ---------------------------------------- 0.0/9.7 MB ? eta -:--:--
   ---------------------------------------- 0.0/9.7 MB ? eta -:--:--
   ---------------------------------------- 0.0/9.7 MB ? eta -:--:--
   ---------------------------------------- 0.0/9.7 MB ? eta -:--:--
   ---------------------------------------- 0.0/9.7 MB ? eta -:--:--
   ---------------------------------------- 0.0/9.7 MB ? eta -:--:--
   ---------------------------------------- 0.0/9.7 MB ? eta -:--:--
   ---------------------------------------- 0.0/9.7 MB ? eta -:--:--
   ---------------------------------------- 0.0/9.7 MB ? eta -:--:--
   ---------------------------------------- 0.0/9.7 MB ? eta -:--:--
   ---------------------------------------- 0.0/9.7 MB ?


[notice] A new release of pip is available: 24.3.1 -> 25.3
[notice] To update, run: python.exe -m pip install --upgrade pip


initiated datasets repo at: C:\Users\SpencerNGUYEN\.pydataset/
Shape of dataset: (153, 6)
Number of unique months: 5


In [5]:
print("Number of rows:", df.shape[0])
print("Number of columns:", df.shape[1])


Number of rows: 153
Number of columns: 6


In [6]:
print("Number of unique months:", df["Month"].nunique())


Number of unique months: 5


**Fill in the blanks.**

The data set has 153 (integer) rows and 6 (integer) columns. There are ___5____ (integer) unique months represented in the data.

#### (b) Create a `Season` column
Create a new column `Season` that takes the value of `Spring` for May, `Summer` for June, July, August, and `Autumn` for September.

Print the first 125 rows of the data.

In [7]:
import numpy as np

df["Season"] = np.where(df["Month"] == 5, "Spring",
                 np.where(df["Month"].isin([6, 7, 8]), "Summer",
                 np.where(df["Month"] == 9, "Autumn", None)))

df.head(125)


,Ozone,Solar.R,Wind,Temp,Month,Day,Season
1,41.0,190.0,7.4,67,5,1,Spring
2,36.0,118.0,8.0,72,5,2,Spring
3,12.0,149.0,12.6,74,5,3,Spring
4,18.0,313.0,11.5,62,5,4,Spring
5,NaN,NaN,14.3,56,5,5,Spring
6,28.0,NaN,14.9,66,5,6,Spring
7,23.0,299.0,8.6,65,5,7,Spring
8,19.0,99.0,13.8,59,5,8,Spring
9,8.0,19.0,20.1,61,5,9,Spring
10,NaN,194.0,8.6,69,5,10,Spring


#### (c) Summary Statistics
Use `describe()` to display summary statistics for temperature in the summer months.

In [8]:
df[df["Season"] == "Summer"]["Temp"].describe()


count    92.000000
mean     82.358696
std       6.285482
min      65.000000
25%      78.000000
50%      82.000000
75%      86.000000
max      97.000000
Name: Temp, dtype: float64

---

## 3. Function to calculate Taxi Fare

Define a function called `get_fare(distance)` that calculates a taxi fare based on the following:
- Base fare is $5.00.
- If distance is greater than 10km, add $1.50 per km for every km **above** 10.
- Return the total fare.

Make use of your function to print 

The fare for 15km is $12.50.

In [9]:
def get_fare(distance):
    base_fare = 5.00
    
    if distance > 10:
        extra_km = distance - 10
        return base_fare + extra_km * 1.50
    else:
        return base_fare

distance = 15
fare = get_fare(distance)

print(f"The fare for {distance}km is ${fare:.2f}.")


The fare for 15km is $12.50.
